tests/csapi/txnid_test.go

```diff
@@ -0,0 +1,178 @@
+	// Conduit scope transaction IDs to the device ID, not the client session.
+	runtime.SkipIf(t, runtime.Conduit)
+
+	deployment := Deploy(t, b.BlueprintCleanHS)
+	defer deployment.Destroy(t)
+
+	deployment.RegisterUser(t, "hs1", "alice", "password", false)
+
+	// Create a first client, which allocates a device ID.
+	c1 := deployment.Client(t, "hs1", "")
+	c1.UserID, c1.AccessToken, c1.DeviceID = c1.LoginUser(t, "alice", "password")
+
+	// Create a room where we can send events.
+	roomID := c1.CreateRoom(t, map[string]interface{}{})
+
+	txnId := 1
+	event := b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "foo",
+		},
+	}
+	// send an event with set txnId
+	eventID1 := c1.SendEventUnsyncedWithTxnID(t, roomID, event, txnId)
+
+	// Create a second client, inheriting the first device ID.
+	c2 := deployment.Client(t, "hs1", "")
+	c2.UserID, c2.AccessToken = c2.LoginUserWithDeviceID(t, "alice", "password", c1.DeviceID)
+	c2.DeviceID = c1.DeviceID
+
+	// send another event with the same txnId
+	eventID2 := c2.SendEventUnsyncedWithTxnID(t, roomID, event, txnId)
+
+	// the two events should have different event IDs as they came from different clients
+	if eventID1 == eventID2 {
+		t.Fatalf("Expected event IDs to be different from two clients sharing the same device ID")
+	}
+}
+
+// TestTxnIdempotency tests that PUT requests idempotency follows required semantics
+func TestTxnIdempotency(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintCleanHS)
+	defer deployment.Destroy(t)
+
+	deployment.RegisterUser(t, "hs1", "alice", "password", false)
+
+	// Create a first client, which allocates a device ID.
+	c1 := deployment.Client(t, "hs1", "")
+	c1.UserID, c1.AccessToken, c1.DeviceID = c1.LoginUser(t, "alice", "password")
+
+	// Create a room where we can send events.
+	roomID1 := c1.CreateRoom(t, map[string]interface{}{})
+	roomID2 := c1.CreateRoom(t, map[string]interface{}{})
+
+	// choose a transaction ID
+	txnId := 1
+	event1 := b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "first",
+		},
+	}
+	event2 := b.Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "second",
+		},
+	}
+
+	// we send the event and get an event ID back
+	eventID1 := c1.SendEventUnsyncedWithTxnID(t, roomID1, event1, txnId)
+
+	// we send the identical event again and should get back the same event ID
+	eventID2 := c1.SendEventUnsyncedWithTxnID(t, roomID1, event1, txnId)
+
+	if eventID1 != eventID2 {
+		t.Fatalf("Expected event IDs to be the same, but they were not")
+	}
+
+	// even if we change the content we should still get back the same event ID as transaction ID is the same
+	eventID3 := c1.SendEventUnsyncedWithTxnID(t, roomID1, event2, txnId)
+
+	if eventID1 != eventID3 {
+		t.Fatalf("Expected event IDs to be the same even with different content, but they were not")
```

Comment:
These error messages are okay but could be more ergonomic. It would be nice to log what the inputs were, in this case `eventID1` and `eventID3`, which you can do by also using an existing heper, in this case `must.EqualStr(got, want, msg)`.